File: Stacks/150. Evaluate Reverse Polish Notation/150. Evaluate Reverse Polish Notation.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for token in tokens:
            if token.lstrip("-").isdigit():
                stack.append(int(token))
            else:
                val2 = stack.pop()
                val1 = stack.pop()
                if token == "+":
                    ans = val1 + val2
                elif token == "-":
                    ans = val1 - val2
                elif token == "*":
                    ans = val1 * val2
                else:
                    ans = int(val1 / val2)
                stack.append(ans)
            
        return stack.pop()

# Time: O(n)
# Space: O(n)

File: Stacks/150. Evaluate Reverse Polish Notation/150. Evaluate Reverse Polish Notation.js
/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function (tokens) {
  let stack = [];

  for (const t of tokens) {
    if (!isNaN(t)) {
      stack.push(Number(t));
    } else {
      const b = stack.pop();
      const a = stack.pop();
      switch (t) {
        case "+":
          stack.push(a + b);
          break;
        case "-":
          stack.push(a - b);
          break;
        case "*":
          stack.push(a * b);
          break;
        case "/":
          stack.push(Math.trunc(a / b));
          break;
        default:
          break;
      }
    }
  }

  return stack.pop();
};

// Time: O(n)
// Space: O(n)
